tls/s2n_handshake_io.c

```diff
@@ -567,6 +582,47 @@ static int s2n_handshake_handle_sslv2(struct s2n_connection *conn)
     return 0;
 }
 
+/* Sometimes the handshake state machine can be blocked on application data. 
+ * For example, if the s2n server is using a remote upstream peer for session cache lookup,
+ * the handshake state machine will be blocked until there is data returned from the 
+ * session cache server. In such a case, there is no data in the underlying io, 
+ * so we can just call the corresponding handler to process the application data.
+ * Note that we need a return value 1 to indicate that app data is not yet avaiable 
+ * and the state machine is still blocked; a return value 0 to indicate the app data is 
+ * successfully processed and state machine advanced; -1 to indicate an error and 
+ * the connection been killed
+ */
+static int handshake_handle_app_data(struct s2n_connection *conn) {
+    int r = ACTIVE_STATE(conn).handler[conn->mode] (conn);
+
+    if (r == 0) {
+        /* if r == 0, then we advance the state machine */
+        GUARD(s2n_stuffer_wipe(&conn->handshake.io));
+        GUARD(s2n_advance_message(conn));
+        goto done;
+    }
+
+    if (r == 1) {
+        /* if r == 1, it means we are still blocked */
+        return r;
+    }
+
+    if (r < 0) {
+        /* if r == -1, we kill the connection */
+        GUARD(s2n_stuffer_wipe(&conn->handshake.io));
+        GUARD(s2n_connection_kill(conn));
+        return r;
+    }
+
+done:    
+    /* We're done with the record, wipe it */
+    //GUARD(s2n_stuffer_wipe(&conn->header_in));
+    //GUARD(s2n_stuffer_wipe(&conn->in));
```

Comment:
Commented out code?

Reply:
removed in https://github.com/songhu/s2n/pull/5